server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java

```diff
@@ -35,11 +55,82 @@ public class RemoteRoutingTableService extends AbstractLifecycleComponent {
     private final Settings settings;
     private final Supplier<RepositoriesService> repositoriesService;
     private BlobStoreRepository blobStoreRepository;
+    private final ThreadPool threadPool;
 
-    public RemoteRoutingTableService(Supplier<RepositoriesService> repositoriesService, Settings settings) {
+    private static final DiffableUtils.NonDiffableValueSerializer<String, IndexRoutingTable> CUSTOM_ROUTING_TABLE_VALUE_SERIALIZER =
+        new DiffableUtils.NonDiffableValueSerializer<String, IndexRoutingTable>() {
+            @Override
+            public void write(IndexRoutingTable value, StreamOutput out) throws IOException {
+                value.writeTo(out);
+            }
+
+            @Override
+            public IndexRoutingTable read(StreamInput in, String key) throws IOException {
+                return IndexRoutingTable.readFrom(in);
+            }
+        };
+
+    public RemoteRoutingTableService(Supplier<RepositoriesService> repositoriesService, Settings settings, ThreadPool threadPool) {
         assert isRemoteRoutingTableEnabled(settings) : "Remote routing table is not enabled";
         this.repositoriesService = repositoriesService;
         this.settings = settings;
+        this.threadPool = threadPool;
+    }
+
+    public CheckedRunnable<IOException> getAsyncIndexMetadataReadAction(
+        String uploadedFilename,
+        Index index,
+        LatchedActionListener<IndexRoutingTable> latchedActionListener) {
+        int idx = uploadedFilename.lastIndexOf("/");
+        String blobFileName = uploadedFilename.substring(idx+1);
+        BlobContainer blobContainer = blobStoreRepository.blobStore().blobContainer(BlobPath.cleanPath().add(uploadedFilename.substring(0,idx)));
+
+        return () -> readAsync(
+            blobContainer,
+            blobFileName,
+            index,
+            threadPool.executor(ThreadPool.Names.GENERIC),
```

Comment:
will prefer a dedicated thread pool

Reply:
Ack, will be using the same threadpool as for the remote state: `REMOTE_STATE_READ`